### PR TITLE
ORC-1024: [C++] Fix inconsistent bloom filter hashing for numeric values

### PR DESCRIPTION
**Pull request by stiga-huang**
### What changes were proposed in this pull request?

This PR fixes inconsistent bloom filter hashing for numeric values between the Java client and C++ client. Before this PR, bloom filters written by Java clients will be incorrectly used in the C++ clients due to the hashing inconsistency. There is an example in the JIRA description.
After this fix, when reading old ORC files that have bloom filters written by old C++ clients, the bloom filters won't be used.

### Why are the changes needed?

In Java codes, hash values are in type "long" which is signed. Bitwise operators like ">>" (shifting) are sign-awared. So we should use "int64_t" instead of "uint64_t" in the corresponding C++ codes.

### How was this patch tested?

Add hashing tests in both Java and C++ codes for numeric values. (TestBloomFilter.java & TestBloomFilter.cc)
Add hashing tests in C++ codes for string values. (TestMurmur3.cc)
Add test for reading an ORC file with bloom filters written by C++ client in version 1.6.11. Verified bad bloom filters won't affect the results. (TestReader.cc)
ORC-1025 will focus on adding more tests.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you again, @stiga-huang . I finished my first round review.

**Comment by stiga-huang** [comment]
I've addressed the comments and updated the JIRA with `Docs Text`.
Thank for your review, @dongjoon-hyun !

**Comment by dongjoon-hyun** [comment]
Merged to main/1.7.

@stiga-huang . Could you make a backporting PR to branch-1.6, please?
There exist some conflicts with branch-1.6.

File: c++/test/TestMurmur3.cc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  TEST(TestMurmur3, testHash64) {
+    uint8_t origin[] = "It was the best of times, it was the worst of times,"
+                       " it was the age of wisdom, it was the age of foolishness,"
+                       " it was the epoch of belief, it was the epoch of incredulity,"
+                       " it was the season of Light, it was the season of Darkness,"
+                       " it was the spring of hope, it was the winter of despair,"
+                       " we had everything before us, we had nothing before us,"
+                       " we were all going direct to Heaven,"
+                       " we were all going direct the other way.";
+    uint32_t len = sizeof(origin) / sizeof(uint8_t) - 1;
```
Comment (dongjoon-hyun) [inline c++/test/TestMurmur3.cc:36]: Unfortunately, this seems to break `CentOS 7` C++ compilation on old C++ versions.
```
[ 75%] Building CXX object c++/test/CMakeFiles/orc-test.dir/TestMurmur3.cc.o
/root/orc/c++/test/TestMurmur3.cc: In member function 'virtual void orc::TestMurmur3_testHash64_Test::TestBody()':
/root/orc/c++/test/TestMurmur3.cc:36:53: error: conversion to 'uint32_t {aka unsigned int}' from 'long unsigned int' may alter its value [-Werror=conversion]
     uint32_t len = sizeof(origin) / sizeof(uint8_t) - 1;
                                                     ^
cc1plus: all warnings being treated as errors
make[2]: *** [c++/test/CMakeFiles/orc-test.dir/TestMurmur3.cc.o] Error 1
make[1]: *** [c++/test/CMakeFiles/orc-test.dir/all] Error 2
make: *** [all] Error 2
FAILED centos7
```

Comment (dongjoon-hyun) [reply c++/test/TestMurmur3.cc:36]: https://issues.apache.org/jira/browse/ORC-1043 is filed.

**Pull request merged**